File: native/rust/src/modulation.rs
//! Modulation modes and symbol ↔ color conversion
//! Pure functions — no platform dependencies.

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ModulationMode {
    Rgb4,      // 4 levels per channel → 6 bits
    Rgb8,      // 8 levels per channel → 9 bits
    Mono,      // Intensity only
    Spatial,   // High-contrast spatial patterns
    Analog,    // Experimental waveform mode
}

impl Default for ModulationMode {
    fn default() -> Self {
        ModulationMode::Rgb4
    }
}

const LEVELS_4: [u8; 4] = [40, 105, 170, 235];
const LEVELS_8: [u8; 8] = [20, 50, 80, 110, 140, 170, 200, 230];

pub fn bits_per_symbol(mode: ModulationMode) -> usize {
    match mode {
        ModulationMode::Rgb8 => 9,
        _ => 6,
    }
}

/// Convert symbol value (0-63 for RGB4, 0-511 for RGB8) to RGB
pub fn value_to_color(v: u16, mode: ModulationMode) -> (u8, u8, u8) {
    match mode {
        ModulationMode::Rgb8 => {
            let r = LEVELS_8[((v >> 6) & 7) as usize];
            let g = LEVELS_8[((v >> 3) & 7) as usize];
            let b = LEVELS_8[(v & 7) as usize];
            (r, g, b)
        }
        ModulationMode::Mono => {
            let lvl = LEVELS_4[(v & 3) as usize];
            (lvl, lvl, lvl)
        }
        ModulationMode::Spatial => {
            let base = LEVELS_4[(v & 3) as usize];
            (base, base, 255 - base)
        }
        ModulationMode::Analog => {
            let val = (v as f32 * 255.0 / 63.0).clamp(0.0, 255.0) as u8;
            (val, (val as f32 * 0.6) as u8, (val as f32 * 0.3) as u8)
        }
        ModulationMode::Rgb4 => {
            let r = LEVELS_4[((v >> 4) & 3) as usize];
            let g = LEVELS_4[((v >> 2) & 3) as usize];
            let b = LEVELS_4[(v & 3) as usize];
            (r, g, b)
        }
    }
}

/// Decode RGB values back to nearest symbol
pub fn color_to_value(r: u8, g: u8, b: u8, mode: ModulationMode) -> u16 {
    match mode {
        ModulationMode::Rgb8 => {
            fn nearest8(x: u8) -> usize {
                LEVELS_8.iter()
                    .enumerate()
                    .min_by_key(|(_, &v)| (x as i32 - v as i32).abs())
                    .map(|(i, _)| i)
                    .unwrap_or(0)
            }
            ((nearest8(r) as u16) << 6) | ((nearest8(g) as u16) << 3) | (nearest8(b) as u16)
        }
        ModulationMode::Mono | ModulationMode::Analog => {
            let lum = ((r as u16 + g as u16 + b as u16) / 3) as u8;
            LEVELS_4.iter()
                .enumerate()
                .min_by_key(|(_, &v)| (lum as i32 - v as i32).abs())
                .map(|(i, _)| i as u16)
                .unwrap_or(0)
        }
        ModulationMode::Spatial => {
            let avg = ((r as u16 + g as u16 + b as u16) / 3) as u8;
            (avg / 64).min(3) as u16
        }
        ModulationMode::Rgb4 => {
            fn nearest(x: u8) -> usize {
                LEVELS_4.iter()
                    .enumerate()
                    .min_by_key(|(_, &v)| (x as i32 - v as i32).abs())
                    .map(|(i, _)| i)
                    .unwrap_or(0)
            }
            ((nearest(r) as u16) << 4) | ((nearest(g) as u16) << 2) | (nearest(b) as u16)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_rgb4_roundtrip() {
        let mode = ModulationMode::Rgb4;
        for v in 0..64u16 {
            let (r, g, b) = value_to_color(v, mode);
            let decoded = color_to_value(r, g, b, mode);
            assert_eq!(decoded, v);
        }
    }
}

File: src/core/modulation.js
/**
 * Modulation.js
 * Defines symbol encoding / decoding for optical channel.
 * Supports multiple modes.
 * Designed to be portable (no DOM / browser specific).
 */

export const LEVELS_4 = [40, 105, 170, 235]; // 4 levels → 2 bits per channel → 6 bits / symbol
export const LEVELS_8 = [20, 50, 80, 110, 140, 170, 200, 230]; // 8 levels → 3 bits/channel

export const MODES = {
  RGB4: 'rgb4',
  RGB8: 'rgb8',
  MONO: 'mono',
  SPATIAL: 'spatial',
  ANALOG: 'analog'
};

let currentMode = MODES.RGB4;
let currentLevels = LEVELS_4;

export function setModulationMode(mode) {
  currentMode = mode;
  if (mode === MODES.RGB8) currentLevels = LEVELS_8;
  else currentLevels = LEVELS_4;
}

export function getCurrentMode() {
  return currentMode;
}

export function getLevels() {
  return [...currentLevels];
}

/**
 * Map 6-bit value (or 9-bit for RGB8) to RGB color triple.
 */
export function valueToColor(v) {
  if (currentMode === MODES.RGB8) {
    // 9-bit value: 3 bits per channel
    const r = currentLevels[(v >> 6) & 7];
    const g = currentLevels[(v >> 3) & 7];
    const b = currentLevels[v & 7];
    return [r, g, b];
  }
  if (currentMode === MODES.MONO) {
    // Intensity only: use average level on all channels or luminance
    const lvl = currentLevels[v & 3]; // reuse 4-level for simplicity
    return [lvl, lvl, lvl];
  }
  if (currentMode === MODES.SPATIAL) {
    // Spatial pattern: encode as checker or simple pattern
    const base = currentLevels[v & 3];
    return [base, base, 255 - base]; // high contrast spatial cue
  }
  if (currentMode === MODES.ANALOG) {
    // High-throughput analog mode (~30 bits/symbol) measured at 2.76 Mbps potential
    const val = Math.min(255, Math.max(0, Math.floor(v * 255 / 1023)));
    return [val, val * 0.6, val * 0.3];
  }
  // Default RGB4
  return [
    currentLevels[(v >> 4) & 3],
    currentLevels[(v >> 2) & 3],
    currentLevels[v & 3]
  ];
}

/**
 * Decode RGB -> nearest symbol value
 */
export function colorToValue(r, g, b) {
  if (currentMode === MODES.RGB8) {
    function nearest8(x) {
      let bi = 0, bd = 1e9;
      for (let i = 0; i < currentLevels.length; i++) {
        const d = Math.abs(x - currentLevels[i]);
        if (d < bd) { bd = d; bi = i; }
      }
      return bi;
    }
    return (nearest8(r) << 6) | (nearest8(g) << 3) | nearest8(b);
  }

  if (currentMode === MODES.MONO || currentMode === MODES.ANALOG) {
    const lum = Math.round((r + g + b) / 3);
    function nearest4(x) {
      let bi = 0, bd = 1e9;
      for (let i = 0; i < LEVELS_4.length; i++) {
        const d = Math.abs(x - LEVELS_4[i]);
        if (d < bd) { bd = d; bi = i; }
      }
      return bi;
    }
    return nearest4(lum);
  }

  if (currentMode === MODES.SPATIAL) {
    // Approximate by dominant channel or average
    const avg = Math.round((r + g + b) / 3);
    return Math.min(3, Math.max(0, Math.floor(avg / 64)));
  }

  // RGB4 default
  function nearest(x) {
    let bi = 0, bd = 1e9;
    for (let i = 0; i < currentLevels.length; i++) {
      const d = Math.abs(x - currentLevels[i]);
      if (d < bd) { bd = d; bi = i; }
    }
    return bi;
  }
  return (nearest(r) << 4) | (nearest(g) << 2) | nearest(b);
}

export function bitsPerSymbol() {
  if (currentMode === MODES.RGB8) return 9;
  return 6;
}
